active-rfcs/0000-v-model-api-change.md

```diff
@@ -0,0 +1,152 @@
+- Start Date: 2019-04-09
+- Target Major Version: 3.x
+- Reference Issues: vuejs/rfcs#8
+- Implementation PR: N/A
+
+# Summary
+
+Adjust `v-model` API when used on custom components.
+
+This builds on top of #8 (Replace `v-bind`'s `.sync` with a `v-model` argument).
+
+# Basic example
+
+# Motivation
+
+Previously, `v-model="foo"` on components roughly compiles to the following:
+
+``` js
+h(Comp, {
+  value: foo,
+  onInput: value => {
+    foo = value
+  }
+})
+```
+
+However, this requires the component to always use the `value` prop for binding with `v-model` when the component may want to expose the `value` prop for a different purpose.
+
+In 2.2 we introduced the `model` component option that allows the component to customize the prop and event to use for `v-model`. However, this still only allows one `v-model` to be used on the component. In practice we are seeing some components that need to sync multiple values, and the other values have to use `v-bind.sync`. We noticed that `v-model` and `v-bind.sync` are fundamentally doing the same thing and can be combined into a single construct by allowing `v-model` to accept arguments (as proposed in #8).
+
+# Detailed design
+
+In 3.0, the `model` option will be removed. `v-model="foo"` (without argument) on a component compiles to the following instead:
+
+``` js
+h(Comp, {
+  modelValue: foo,
+  'onUpdate:modelValue': value => {
+    foo = value
+  }
+})
+```
+
+If the component wants to support `v-model` without an argument, it should expect a prop named `modelValue`. To sync its value back to the parent, the child should emit an event named `"update:modelValue"` (see [Render Function API change](https://github.com/vuejs/rfcs/blob/render-fn-api-change/active-rfcs/0000-render-function-api-change.md) for details on the new VNode data structure).
+
+The default compilation output prefixes the prop and event names with `model` to avoid conflict with common prop names.
+
+RFC #8 proposes the ability for `v-model` to accept arguments. The argument can be used to denote the prop `v-model` should bind to. `v-model:value="foo"` compiles to:
+
+``` js
+h(Comp, {
+  value: foo,
+  'onUpdate:value': value => {
+    foo = value
+  }
+})
+```
+
+In this case, the child component expects a `value` prop and emits `"update:value"` to sync.
+
+Note that this enables multiple `v-model` bindings on the same component, each syncing a different prop, without the need for extra options in the component:
+
+``` html
+<InviteeForm
+  v-model:name="inviteeName"
+  v-model:email="inviteeEmail"
+/>
+```
+
+## Handling Modifiers
+
+In 2.x, we have hard-coded support for modifiers like `.trim` on component `v-model`. However, it would be more useful if the component can support custom modfiers. In v3, modifiers added to a componen `v-model` will be provided to the component via the `modelModifiers` prop:
```

Comment:
```suggestion
In 2.x, we have hard-coded support for modifiers like `.trim` on component `v-model`. However, it would be more useful if the component can support custom modfiers. In v3, modifiers added to a component `v-model` will be provided to the component via the `modelModifiers` prop:
```